### PR TITLE
Add boundary feature

### LINKED ISSUE #34: Truck boundaries limit
Hello! Thank you for putting the effort to have this. It is great.

I've been trying to modify it a bit in order to implement bound limits to the **truck** function, but I couldn't find a "nice" way to put it together. I always mess up with something.
 
I am working in a map that needs to be panned within the boundaries oif the model,  but nothing I do does the purpose.

Any tip or advice?

### PR DESCRIPTION
**Pull request by FMS-Cat**
An attempt to resolve #34

- New property: `.boundary`
  - Is a `THREE.Box3` encloses the camera target into a certain region
- New property: `.boundaryFriction`
  - Friction ratio of the boundary, can tweak boundaries behavior when the camera target is in collision
- New property: `.boundaryEnclosesCamera`
  - Whether camera position also should be enclosed in the boundary or not
- plus
  - Update `examples/basic.html`
  - Update README

### CODE REVIEW

File: src/camera-controls.js [hunk 1]
```diff
@@ -882,6 +902,45 @@ export default class CameraControls extends EventDispatcher {
 
 	}
 
+	_encloseToBoundary( position, offset, friction ) {
```
Comment (yomotsu) [inline src/camera-controls.js]: Do we need the third argument? we could use this. `boundaryFriction` in the method.
```suggestion
	_encloseToBoundary( position, offset ) {
```

File: src/camera-controls.js [hunk 2]
```diff
@@ -882,6 +902,45 @@ export default class CameraControls extends EventDispatcher {
...
+		if ( offsetLength === 0.0 ) { // sanity check
+
+			return position;
+
+		}
+
+		const t1 = _v3B.copy( offset ).add( position ); // target
+		const tc = this.boundary.clampPoint( t1, _v3C ); // clamped target
+		const dtc = tc.sub( t1 ); // t1 -> tc
+		const dtcl = dtc.length(); // length of dtc
```
Comment (yomotsu) [inline src/camera-controls.js]: Full-form name is preferable rather than abbreviated.

**Comment by yomotsu** [comment]
I sent an additonal PR https://github.com/yomotsu/camera-controls/pull/37
Could you take a look that please?

File: src/camera-controls.js [hunk 3]
```diff
@@ -804,6 +813,17 @@ export default class CameraControls extends EventDispatcher {
 		this.object.position.setFromSpherical( this._spherical ).add( this._target );
 		this.object.lookAt( this._target );
 
+		if ( this.boundaryEnclosesCamera ) {
+
+			this._encloseToBoundary(
```
Comment (yomotsu) [inline src/camera-controls.js]: I think just `this.boundary.clampPoint` should be enough at here
```diff
-			this._encloseToBoundary(
-				this.object.position.copy( this._target ),
-				_v3A.setFromSpherical( this._spherical ),
-				1.0
-			);
+			this.boundary.clampPoint(
+				this.object.position,
+				this.object.position
+			);
```

File: src/camera-controls.js (hunk 2 above)
Comment (FMS-Cat) [reply src/camera-controls.js]: I used mathematic oriented notation here, and also scope is pretty short so I thought it's fine,
but if I need to rename these, I'd choose `newTarget` , `clampedTarget` , `deltaClampedTarget` , `deltaClampedTargetLength`

File: src/camera-controls.js (hunk 3 above)
Comment (FMS-Cat) [reply src/camera-controls.js]: I prefer the current one. I don't like how the camera moves on the edge when its "friction" is zero

...also I kinda want to change the term "friction" ? How we should call him

File: src/camera-controls.js (hunk 2 above)
Comment (yomotsu) [reply src/camera-controls.js]: That makes sense. But if you could rename to those, should be better for other users who may read the code🙏

File: src/camera-controls.js (hunk 3 above)
Comment (yomotsu) [reply src/camera-controls.js]: Sure keep it as is then👍

File: src/camera-controls.js (hunk 2 above)
Comment (FMS-Cat) [reply src/camera-controls.js]: Okay 'n' done

**Comment by FMS-Cat** [comment]
Types!!! I forgot that, thanks!

File: src/camera-controls.js (hunk 2 above)
Comment (yomotsu) [reply src/camera-controls.js]: Thanks a lot!

**Comment by yomotsu** [comment]
I guess we should switch to TS entirely sometime soon.

File: src/camera-controls.js [hunk 4]
```diff
@@ -694,6 +707,17 @@ export default class CameraControls extends EventDispatcher {
 
 	}
 
+	setBoundary( box3 ) {
+
+		this._boundary.copy( box3 || new THREE.Box3(
```
Comment (FMS-Cat) [inline src/camera-controls.js]: How about this one?
Added default property for the `setBoundary` method

Comment (yomotsu) [reply src/camera-controls.js]: Looks good!

File: src/camera-controls.js [hunk 5]
```diff
@@ -882,6 +914,47 @@ export default class CameraControls extends EventDispatcher {
...
+			return position;
+
+		}
+
+		 // See: https://twitter.com/FMS_Cat/status/1106508958640988161
+
+		const newTarget = _v3B.copy( offset ).add( position ); // target
+		const clampedTarget = this.boundary.clampPoint( newTarget, _v3C ); // clamped target
+		const deltaClampedTarget = clampedTarget.sub( newTarget ); // newTarget -> clampedTarget
+		const deltaClampedTargetLength2 = deltaClampedTarget.lengthSq(); // squared length of deltaClampedTarget
```
Comment (yomotsu) [inline src/camera-controls.js]: Nice found! this should much faster than `length()`!

**Comment by yomotsu** [comment]
merged! many thanks.

**Pull request merged**